be/src/storage/persistent_index.cpp

```diff
@@ -613,6 +629,68 @@ class FixedMutableIndex : public MutableIndex {
         return Status::OK();
     }
 
+    Status load(const MutableIndexMetaPB& meta) {
+        IndexSnapshotMetaPB snapshot_meta = meta.snapshot();
+        EditVersion start_version = snapshot_meta.version();
+        PagePointerPB page_pb = snapshot_meta.data();
+        size_t snapshot_off = page_pb.offset();
+        size_t snapshot_size = page_pb.size();
+
+        std::string index_file_name = get_l0_index_file_name(_path, start_version);
+        std::shared_ptr<FileSystem> fs;
+        ASSIGN_OR_RETURN(fs, FileSystem::CreateSharedFromString(_path));
+
+        // Assuming that the snapshot is always at the beginning of index file,
+        // if not, we can't call phmap.load() directly because phmap.load() alaways
+        // reads the contents of the file from the beginning
+        phmap::BinaryInputArchive ar_in(index_file_name.data());
+        if (snapshot_size > 0 && !_map.load(ar_in)) {
+            std::string err_msg = strings::Substitute("failed load snapshot from file $0", index_file_name);
+            LOG(WARNING) << err_msg;
+            return Status::InternalError(err_msg);
+        }
+        ASSIGN_OR_RETURN(auto read_file, fs->new_random_access_file(index_file_name));
+        // if mutable index is empty, set _offset as 0, otherwise set _offset as snapshot size
+        _offset = snapshot_off + snapshot_size;
+        int n = meta.wals_size();
+        // read wals and build hash map
+        for (int i = 0; i < n; i++) {
+            const auto& wal_pb = meta.wals(i);
+            const auto& page_pb = wal_pb.data();
+            size_t offset = page_pb.offset();
+            size_t size = page_pb.size();
+            _offset = offset + size;
+            size_t kv_size = KeySize + sizeof(IndexValue);
+            size_t nums = size / kv_size;
+            std::string buff;
+            while (nums > 0) {
+                size_t batch_num = (nums > 4096) ? 4096 : nums;
+                raw::stl_string_resize_uninitialized(&buff, batch_num * kv_size);
+                RETURN_IF_ERROR(read_file->read_at_fully(offset, buff.data(), buff.size()));
+                uint8_t keys[KeySize * batch_num];
+                std::vector<IndexValue> values;
+                values.reserve(batch_num);
+                size_t buf_offset = 0;
+                for (size_t j = 0; j < batch_num; ++j) {
+                    memcpy(keys + j * KeySize, buff.data() + buf_offset, KeySize);
```

Comment:
If KeySize is not very large, you can use memcpy_inlined instead

Reply:
> If KeySize is not very large, you can use memcpy_inlined instead

done, change in https://github.com/StarRocks/starrocks/pull/8866/commits/3e3638539832e1ffbda9f695c26ae439bed936c9